src/backend/joanie/payment/models.py

```diff
@@ -1 +1,463 @@
+                "type": self.type,
+            },
+            "order": {
+                "amount": {
+                    "currency": currency,
+                    "subtotal": str(net_amount),
+                    "total": str(total_amount),
+                    "vat_amount": str(vat_amount),
+                    "vat": str(vat),
+                },
+                "company": settings.JOANIE_INVOICE_COMPANY_CONTEXT,
+                "customer": {
+                    "address": self.recipient_address,
+                    "name": self.recipient_name,
+                },
+                "seller": {
+                    "address": settings.JOANIE_INVOICE_SELLER_ADDRESS,
+                },
+            },
+        }
+
+        try:
+            localized_context = self.localized_context[language_code]
+        except KeyError:
+            fallback_language = settings.PARLER_DEFAULT_LANGUAGE_CODE
+            localized_context = self.localized_context[fallback_language]
+
+        return merge_dict(base_context, localized_context)
+
+    def normalize_reference(self):
+        """
+        Generate a normalized reference related to the date
+        and the related order
+        """
+        order_uid_fragment = str(self.get_related_order().uid).split("-", maxsplit=1)[0]
+        timestamp = int(timezone.now().timestamp() * 1_000)  # Time in milliseconds
+
+        return f"{order_uid_fragment}-{timestamp}"
+
+    def clean(self):
+        if (self.order is None and self.parent is None) or (self.order and self.parent):
+            raise ValidationError(_("Invoice must have an order or a parent."))
+
+        if self.parent:
+            self.recipient_name = self.parent.recipient_name
+            self.recipient_address = self.parent.recipient_address
+
+        if self.type == payment_enums.INVOICE_TYPE_CREDIT_NOTE:
+            if (
+                self.amount.amount * -1  # pylint: disable=no-member
+                > self.parent.amount.amount  # pylint: disable=no-member
+            ):
+                raise ValidationError(
+                    _(
+                        "Credit note amount cannot be greater than its related invoice amount."
+                    )
+                )
+
+        if self.pk is None:
+            self.reference = self.normalize_reference()
+
+        return super().clean()
+
+    def save(self, *args, **kwargs):
+        """
+        Enforce validation each time an instance is saved.
+
+        On creation, we also create a context for each active languages.
+        """
+
+        is_new = self.pk is None
+
+        if is_new:
+            self._update_context()
+
+        self.full_clean()
+        super().save(*args, **kwargs)
+
+
+class Transaction(models.Model):
+    """
+    Transaction model represents financial transactions
+    (debit, credit) related to an invoice.
+    """
+
+    reference = models.CharField(db_index=True, max_length=40, unique=True)
```

Comment:
editable False?

Reply:
Yes as this field is normalized on creation 👍

Reply:
Unlike invoice model,  transaction model does not normalize `reference` field on creation.
So to register a transaction manually, a staff user should specify a reference. I think it's good idea to not normalize this field mainly because we are currently using the payment provider payment/refund id as reference when Joanie is notified. In this way, with the transaction reference we are able to find the related payment from the payment provider admininistration.

Reply:
I would say that transactions do not need any reference field... unless you need to refer to an external reference like you do with PayPlug...
Allowing manual references put uniqueness at risk? What would happen if, on a PayPlug call, the reference proposed by PayPlug was already taken by a manual transaction?

Reply:
Yes it makes sense. Reference could be an optional field to register external reference!

Reply:
+1 and editable False